File: apps/api/src/akp_api/lib/crypto.py
"""Password and opaque-token crypto — parity with apps/api-node/src/lib/crypto.ts."""

from __future__ import annotations

import hashlib
import hmac
import secrets

from argon2 import PasswordHasher
from argon2.exceptions import VerifyMismatchError


def hash_password(password: str, memory_cost: int) -> str:
    hasher = PasswordHasher(time_cost=3, memory_cost=memory_cost, parallelism=1)
    return hasher.hash(password)


def verify_password(password_hash: str, password: str) -> bool:
    hasher = PasswordHasher()
    try:
        return hasher.verify(password_hash, password)
    except (VerifyMismatchError, Exception):
        return False


def generate_opaque_token(nbytes: int = 32) -> str:
    return secrets.token_urlsafe(nbytes)


def hash_token(token: str) -> str:
    return hashlib.sha256(token.encode("utf-8")).hexdigest()


def safe_compare_hex(a: str, b: str) -> bool:
    try:
        return hmac.compare_digest(bytes.fromhex(a), bytes.fromhex(b))
    except ValueError:
        return False

File: apps/api-node/src/lib/crypto.ts
import { createHash, randomBytes, timingSafeEqual } from 'node:crypto';
import argon2 from 'argon2';

/**
 * Cryptographic helpers for the auth subsystem.
 *
 * - Passwords: Argon2id (memory-hard, resistant to GPU cracking).
 * - Opaque tokens (refresh tokens, API keys, invites): high-entropy random
 *   strings hashed with SHA-256 before storage. We only ever persist the hash,
 *   so a database leak does not expose usable credentials.
 */

const ARGON2_TYPE = argon2.argon2id;

export async function hashPassword(password: string, memoryCost: number): Promise<string> {
  return argon2.hash(password, {
    type: ARGON2_TYPE,
    memoryCost,
    timeCost: 3,
    parallelism: 1,
  });
}

export async function verifyPassword(hash: string, password: string): Promise<boolean> {
  try {
    return await argon2.verify(hash, password);
  } catch {
    // A malformed hash should never authenticate a user.
    return false;
  }
}

/** Generate a URL-safe, high-entropy opaque token (default 256 bits). */
export function generateOpaqueToken(bytes = 32): string {
  return randomBytes(bytes).toString('base64url');
}

/** Deterministically hash an opaque token for storage/lookup. */
export function hashToken(token: string): string {
  return createHash('sha256').update(token).digest('hex');
}

/** Constant-time comparison of two hex-encoded hashes. */
export function safeCompareHex(a: string, b: string): boolean {
  const bufA = Buffer.from(a, 'hex');
  const bufB = Buffer.from(b, 'hex');
  if (bufA.length !== bufB.length) return false;
  return timingSafeEqual(bufA, bufB);
}
